src/nvim/edit.c

```diff
@@ -1352,24 +1352,21 @@ ins_redraw (
   if (char_avail())
     return;
 
-  /* Trigger CursorMoved if the cursor moved.  Not when the popup menu is
-   * visible, the command might delete it. */
-  if (ready && (
-        has_cursormovedI()
-        ||
-        curwin->w_p_cole > 0
-        )
+  // Trigger CursorMoved if the cursor moved.  Not when the popup menu is
+  // visible, the command might delete it.
+  if (ready && (has_event(EVENT_CURSORMOVEDI) || curwin->w_p_cole > 0)
       && !equalpos(last_cursormoved, curwin->w_cursor)
-      && !pum_visible()
-      ) {
-    /* Need to update the screen first, to make sure syntax
-     * highlighting is correct after making a change (e.g., inserting
-     * a "(".  The autocommand may also require a redraw, so it's done
-     * again below, unfortunately. */
-    if (syntax_present(curwin) && must_redraw)
+      && !pum_visible()) {
+    // Need to update the screen first, to make sure syntax
+    // highlighting is correct after making a change (e.g., inserting
+    // a "(".  The autocommand may also require a redraw, so it's done
+    // again below, unfortunately.
+    if (syntax_present(curwin) && must_redraw) {
       update_screen(0);
-    if (has_cursormovedI())
-      apply_autocmds(EVENT_CURSORMOVEDI, NULL, NULL, FALSE, curbuf);
+    }
+    if (has_event(EVENT_CURSORMOVEDI)) {
```

Comment:
Let's leave it like this for now, but I wonder what (imagined?) performance problems `apply_autocmds` have that made this extra check necessary.